### ISSUE TITLE
Flow infers wrong type for disjoint union sentinels.

### ISSUE THREAD

**Issue by mkscrg** [facebook/flow#5116]
Flow seems to infer the wrong type for a disjoint union discriminator. [TryFlow example](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AQQMYoCWsAdqALygDeo62AXKAOQBiA8u8wDSiSywmpAK4BbAEZYATqAC+oAD406mLE2YAhXACUeocQEMpTcQOhyA3MhB0AFkQDOoIqQBuB6EQAmofLFEYRo5koA62sMLQPvgGwg44BqCscPCg0lKwxsh+pA4ovv6BUlgAIkTFhJSgABQGhEwExGQAlEyusN6UAHw0iKDOkDV1KAB09DgUkyxaAFrMzcqgNmERUWlSGVIAhKCzoICQRHiEJKRjqn39Bbmw0FgjcADm1cz8wqQ+ibPzVv2yiLJWRA2FzuTzRQpBBwhEKJQzFUIoKQuB6gN4nXyxeJORLJBDrTa8AxOLAAD2whCwXgYyHGRyapAAKqoqmxOMxFNMdMwrDk8gUAkYsABlRHIgCqpHRVFq4waxzITOwrVA7U6FB61AuREGMuZkyomlwcwWmsuVyht3usCeLwi71AnyN3wufwB1jA8BwMXIcRwABIAAoZbBSdCKrAAHh6KFgfEy+C9mQq+UktgMrhIMngRBQtjpJzO2CBYFzjgxvuxSRS+MyhOJZKwFKp2TIfL8AuKQdgIbDeqG9XzLTaHR86t6-V5+WG4aYgeD0l72AjjROvAAROM1z0qMNC1gfgN+yhw5QpobjePLrybndHs9XvbHcaD66LEA)

```js
type Action = { type: 'FOO', foo: number } | { type: 'BAR', bar: bool };
```

For any `act: Action`, it should infer `act.type` as the union of the sentinel values `'FOO' | 'BAR'`. However it seems to infer some other type, causing wrong behavior.

This function `compare` should generate a Flow error:

```js
const compare = (act: Action): void => {
  if (act.type === 'BAZ') {  // should error! BAZ ∉ Action.type
    console.log('found a BAZ');
  }
};
```

Specifically, it should generate the same error as if we'd explicitly defined the sentinel type:

```js
type ActionType = 'FOO' | 'BAR';
const compare = (atype: ActionType): void => {
  if (atype === 'BAZ') {
    console.log('found a BAZ');
  }
};

// 3:   if (atype === 'BAZ') {
//                    ^ string literal `BAZ`. This type is incompatible with
// 2: const compare = (atype: ActionType): void => {
//                            ^ string enum
```

This is the root issue behind #4893, #5083, and likely others.

#### Workaround

We can force correct type checking for disjoint unions by using `$PropertyType<Action, "type">`. The code and the resulting error are much more verbose, and we have to remember to use it every time, but hey it works ¯\\_(ツ)_/¯

```js
const compare = (act: Action): void => {
  const actType: $PropertyType<Action, "type"> = act.type;
  if (actType === 'BAZ') {
    console.log('found a BAZ');
  }
};

// 3:   if (actType === 'BAZ') {
//                      ^ string literal `BAZ`. This type is incompatible with
// 2:   const actType: $PropertyType<Action, "type"> = act.type;
//                     ^ union: property type(s)
// Member 1:
// 2:   const actType: $PropertyType<Action, "type"> = act.type;
//                     ^ property typeError:
// 3:   if (actType === 'BAZ') {
//                      ^ string literal `BAZ`. This === check always fails because `BAZ` is not the same string as `FOO`
// 2:   const actType: $PropertyType<Action, "type"> = act.type;
//                     ^ string literal `FOO`
// Member 2:
// 2:   const actType: $PropertyType<Action, "type"> = act.type;
//                     ^ property typeError:
// 3:   if (actType === 'BAZ') {
//                      ^ string literal `BAZ`. This === check always fails because `BAZ` is not the same string as `BAR`
// 2:   const actType: $PropertyType<Action, "type"> = act.type;
//                     ^ string literal `BAR`
```

**Comment by MarcCoquand** [comment]
This problem also seem to affect switch cases. 

```
export type Msg = 
    | {type: 'NoOp', data: void}
    | {type: 'SetText', data: string}

const update = (mdl: Model, msg: Msg): Model => {
    const msgType: $PropertyType<Msg, "type"> = msg.type;

    switch(msgType) {

        case 'NoOp':
            return mdl;

        case 'SetText':
            const newModel = merge(mdl, {text: msg.data});
            return newModel;

        //Should throw error or at least a warning? 
        case 'Uggachaka':
            return mdl;

        default:
            throw 'error';
}
```
Worth noting though that
```
const update = (mdl: Model, msg: Msg): Model => {
    const msgType: $PropertyType<Msg, "type"> = msg.type;

    switch(msgType) {

        case 'NoOp':
            return mdl;

        //With this order throws an error 
        case 'Uggachaka':
            return mdl;

        case 'SetText':
            const newModel = merge(mdl, {text: msg.data});
            return newModel;

        default:
            throw 'error';
}
```

Throws an error. So that's good at least. 

I also found that the flow doesn't throw any errors with tuples so if you have:

```
type Msg = 
    | ['NoOp', void]
    | ['SetText', string]

const update = (mdl: Model, msg: Msg): Model => {
    switch(msg[0]) {

        case 'NoOp':
            return mdl;

        //Should throw error but doesn't 
        case 'Uggachaka':
            messupeverything();
            return mdl;

        case 'SetText':
            const newModel = merge(mdl, {text: msg[1]});
            return newModel;

        default:
            throw 'error';
}
```
we have the same problem as well.

**Comment by mkscrg** [comment]
Fixed by Flow 0.58! 🎉🎉🎉
